Enhance Noteworthy Differences to Python in Documentation (#35268)
JuliaLang/julia: doc/src/manual/noteworthy-differences.md

## Noteworthy differences from Python

  * Julia's `for`, `if`, `while`, etc. blocks are terminated by the `end` keyword. Indentation level
    is not significant as it is in Python. Unlike Python, Julia has no `pass` keyword.
  * Strings are denoted by double quotation marks (`"text"`) in Julia (with three double quotation marks for multi-line strings), whereas in Python they can be denoted either by single (`'text'`) or double quotation marks (`"text"`). Single quotation marks are used for characters in Julia (`'c'`).

>   * String concatination is done with `*` in Julia, not `+` like in Python. Analogously, string repetition is done with `^`, not `*`.

Reviewer: ```suggestion
  * String concatenation is done with `*` in Julia, not `+` like in Python. Analogously, string repetition is done with `^`, not `*`.
```

>   * String concatenation is done with `*` in Julia, not `+` like in Python. Analogously, string repetition is done with `^`, not `*`.

Reviewer 2: Maybe talk about "implicit concatenation" as in `"a" "b"` in Python does not exist in Julia?  (Maybe it's not the technical name.)

Author: I was not aware of this Python behavior until now (but probably it is not best practice to use it anyhow). I'll include this point.

>   * Python Lists correspond to Julia Arrays of type Any (flexible, but slow). Numpy Arrays correspond to typed Julia Arrays.

Reviewer 2: ```suggestion
  * Python Lists correspond to Julia `Vector{Any}` (flexible, but slow). Numpy arrays correspond to typed Julia Arrays.
```

Also, "Numpy arrays correspond to typed Julia Arrays" is kind of not technically true as `dtype` can be `object`?

Author: Suggestion: 

> Numpy arrays (with concrete data types, not `object`) correspond...

Reviewer 2: I'm not sure if Python programmers are familiar with the notion of concrete type.  Also, even in Julia, array of concrete non-isbits objects can be as slow as `Array{Any}`.

So maybe something like this?

```suggestion
  * Python Lists correspond to Julia `Vector{Any}` (flexible, but slow). Machine-native arrays such as Numpy arrays can be represented by built-in type `Array` in Julia.
```

Not sure if "Machine-native arrays" is the right word though. I just wanted to say `isbits` element type without saying it. Maybe just "Fast arrays?"

Author: What about this formulation?

> Fast arrays such as Numpy arrays with machine-native data types (like `np.int64`, `np.float64`) can be represented by the built-in type `Array` with fixed `isbits` element type in Julia.

Reviewer 2: I think the complication is that both Numpy and Julia can store "struct types" but machine-native may sound like "scalar" types like `Float64` (i.e., the data types with corresponding machine register).

Maybe something like this?

> "Fast" arrays such as Numpy arrays that store elements in-place can be represented by ...

Maybe adding examples can help

> ... that store elements in-place (i.e., `dtype` is `np.float64`, `[('f1', np.uint64), ('f2', np.int32)]`, etc.) can be represented by built-in type `Array` in Julia (i.e., `eltype` is a `isbitstype` like `Float64`, `NamedTuple{(:f1, :f2), Tuple{UInt64, Int32}}`, etc.).

Author: This point got more complicated than I initially thought, I was not aware that it is possible to store structs in-place with numpy.
Adding the examples would certainly help imho.

Author: updated using your suggestion

>   * Python Lists correspond to Julia `Vector{Any}` (flexible, but slow). "Fast" arrays such as Numpy arrays that store elements in-place (i.e., `dtype` is `np.float64`, `[('f1', np.uint64), ('f2', np.int32)]`, etc.) can be represented by built-in type `Array` in Julia (i.e., `eltype` is a `isbitstype` like `Float64`, `NamedTuple{(:f1, :f2), Tuple{UInt64, Int32}}`, etc.).

Reviewer 3: ```suggestion
  * Python Lists—flexible but slow—correspond to the Julia `Vector{Any}` type or more generally `Vector{T}` where `T` is some non-concrete element type. "Fast" arrays like Numpy arrays that store elements in-place (i.e., `dtype` is `np.float64`, `[('f1', np.uint64), ('f2', np.int32)]`, etc.) can be represented by `Array{T}` where `T` is a concrete, immutable element type. This includes built-in types like `Float64`, `Int32`, `Int64` but also more complex types like `Tuple{UInt64,Float64}` and many user-defined types as well.
```

>   * Python Lists—flexible but slow—correspond to the Julia `Vector{Any}` type or more generally `Vector{T}` where `T` is some non-concrete element type. "Fast" arrays like Numpy arrays that store elements in-place (i.e., `dtype` is `np.float64`, `[('f1', np.uint64), ('f2', np.int32)]`, etc.) can be represented by `Array{T}` where `T` is a concrete, immutable element type. This includes built-in types like `Float64`, `Int32`, `Int64` but also more complex types like `Tuple{UInt64,Float64}` and many user-defined types as well.

Reviewer: Like so @strickek?

```
* Python Lists -- flexible but slow -- correspond to ...
```

Reviewer: If I remember correctly, in LaTeX Europeans prefer `begin -- some remarks -- end of sentence.` whereas Americans prefer `begin---some remarks---end of sentence.`.

Author: I would leave a space before and after the dash, but I am a European ;-)
In the docs I could find examples for this in `faq.md`, e.g. `...of a given type — that is,...`.
Should we use the same notation here?

  * In Julia, indexing of arrays, strings, etc. is 1-based not 0-based.
  * Julia's slice indexing includes the last element, unlike in Python. `a[2:3]` in Julia is `a[1:3]`
    in Python.
  * Julia does not support negative indices. In particular, the last element of a list or array is
    indexed with `end` in Julia, not `-1` as in Python.

>   * Julia requires end for indexing until the last element. `x[1:]` in Python is equivalent with x[2:end] in Julia.

Reviewer: ```suggestion
  * Julia requires `end` for indexing until the last element. `x[1:]` in Python is equivalent with x[2:end] in Julia.
```

>   * Julia requires `end` for indexing until the last element. `x[1:]` in Python is equivalent with x[2:end] in Julia.

Reviewer 4: ```suggestion
  * Julia requires `end` for indexing until the last element. `x[1:]` in Python is equivalent to `x[2:end]` in Julia.
```

>   * Julia’s range indexing has the format of `x[start:step:stop]`, whereas Python's format is `x[start:stop:step]`. Hence, `x[0:10:2]` in Python is equivalent to `x[1:2:10]` in Julia. Similarly, `x[::-1]` in Python, which refers to the reversed array, is equivalent to `x[end:-1:1]` in Julia.

Reviewer: This is the last occurence in this file:
```suggestion
  * Julia's range indexing has the format of `x[start:step:stop]`, whereas Python's format is `x[start:stop:step]`. Hence, `x[0:10:2]` in Python is equivalent to `x[1:2:10]` in Julia. Similarly, `x[::-1]` in Python, which refers to the reversed array, is equivalent to `x[end:-1:1]` in Julia.
```

Author: I thought I got all of them... Thanks!

>   * Julia's range indexing has the format of `x[start:step:stop]`, whereas Python's format is `x[start:stop:step]`. Hence, `x[0:10:2]` in Python is equivalent to `x[1:2:10]` in Julia. Similarly, `x[::-1]` in Python, which refers to the reversed array, is equivalent to `x[end:-1:1]` in Julia.

Reviewer 3: ```suggestion
  * Julia's range indexing has the format of `x[start:step:stop]`, whereas Python's format is `x[start:(stop+1):step]`. Hence, `x[0:10:2]` in Python is equivalent to `x[1:2:10]` in Julia. Similarly, `x[::-1]` in Python, which refers to the reversed array, is equivalent to `x[end:-1:1]` in Julia.
```

  * The commonly used `Int` type corresponds to the machine integer type (`Int32` or `Int64`).
    This means it will overflow, such that `2^64 == 0`. If you need larger values use another appropriate type,
    such as `Int128`, [`BigInt`](@ref) or a floating point type like `Float64`.
  * The imaginary unit `sqrt(-1)` is represented in Julia as `im`, not `j` as in Python.
  * In Julia, the exponentiation operator is `^`, not `**` as in Python.
  * Julia uses `nothing` of type `Nothing` to represent a null value, whereas Python uses `None` of type `NoneType`.
  * In Julia, the standard operators over a matrix type are matrix operations, whereas, in Python, the standard operators are element-wise operations. When both `A` and `B` are matrices, `A * B` in Julia performs matrix multiplication, not element-wise multiplication as in Python. `A * B` in Julia is equivalent with `A @ B` in Python, whereas `A * B` in Python is equivalent with `A .* B` in Julia.
  * The adjoint operator `'` in Julia returns an adjoint of a vector (a lazy representation of row vector), whereas the transpose operator `.T` over a vector in Python returns the original vector (non-op).

>   * Functions in Python have one implementation (no polymorphism), whereas in Julia a function may contain multiple concrete implementations (called *Methods*), selected via multiple dispatch.

Reviewer 5: As a reader, I have found it helpful to keep this list entirely in the pattern of "Julia behavior in contrast to other language behavior."

Author: I aligned the pattern as you suggested for most of the points.
A few points do not fit so well content-wise into this pattern (e.g. showing how methods of Python objects translate to Julia), therefore I kept them as they are. 
Could you please check if it is fine for you now?

Reviewer 5: Yes, that's awesome, thanks!  I know it's not always possible/readable that way, this is much easier for me to read. 👍

  * There are no classes in Julia. Instead they are structures (mutable or immutable), containing data but no methods.
  * Calling a method of a class in Python (`a = MyClass(x), x.func(y)`) corresponds to a function call in Julia, e.g. `a = MyStruct(x), func(x::MyStruct, y)`. In general, multiple dispatch is more flexible and powerful than the Python class system.
  * Python classes can inherit from one or more superclasses. Julia structures may have exactly one abstract supertype.
  * The Python code structure is defined by directories (Packages) and files (Modules), whereas the logical Julia program structure (Packages and Modules) is independent of the file strucutre (`include` for additional files).
  * Conditional expression in Python `1 if x > 0 else -1` corresponds to the ternary operator `x > 0 ? 1 : -1` in Julia.
  * In Python, the `@` symbol refers to a decorator, whereas in Julia it refers to a macro.

>   * Exception handling in Julia is done using `try` - `catch` - `finally`, instead of `try` - `except` - `finally`. In contrast to Python, it is not recommended to use exception handling as part of the normal workflow in Julia due to performance reasons.

Reviewer: ```suggestion
  * Exception handling in Julia is done using `try`-`catch`-`finally`, instead of `try`-`except`-`finally`. In contrast to Python, it is not recommended to use exception handling as part of the normal workflow in Julia due to performance reasons.
```
to be in line with the last bullet point.

  * In Julia loops are fast, there is no need to write "vectorized" code for performance reasons.
  * Be careful with non-constant global variables in Julia, especially in tight loops. Since you can write close-to-metal code in Julia (unlike Python), the effect of globals can be drastic (see [Performance Tips](@ref man-performance-tips)).

>   * In Python, the majority of values can be used in logical contexts (e.g. `if "a":` means the following block is executed, and `if "":` means it is not). In Julia, you need explicit conversion to `Bool` (e.g. `if "a"` throws an exception).

Reviewer 3: ```suggestion
  * In Python, the majority of values can be used in logical contexts (e.g. `if "a":` means the following block is executed, and `if "":` means it is not). In Julia, you need explicit conversion to `Bool` (e.g. `if "a"` throws an exception). If you want to test for a non-empty string in Julia, you would explicitly write `if !isempty("")`.
```

  * The commonly used `Int` type corresponds to the machine integer type (`Int32` or `Int64`).
    This means it will overflow, such that `2^64 == 0`. If you need larger values use another appropriate type,
    such as `Int128`, [`BigInt`](@ref) or a floating point type like `Float64`.
  * The imaginary unit `sqrt(-1)` is represented in Julia as `im`, not `j` as in Python.
  * In Julia, the exponentiation operator is `^`, not `**` as in Python.

>   * In Julia uses `nothing` of type `Nothing` to represent null value, whereas Python uses `None` of type `NoneType`.

Reviewer 4: ```suggestion
  * Julia uses `nothing` of type `Nothing` to represent a null value, whereas Python uses `None` of type `NoneType`.
```

>   * In Julia, we use `nothing` of type `Nothing` to represent null value, whereas Python uses `None` of type `NoneType`.

Reviewer: ```suggestion
  * In Julia uses `nothing` of type `Nothing` to represent null value, whereas Python uses `None` of type `NoneType`.
```

>   * In Julia, the standard operators over a matrix type are matrix operations, whereas, in Python, the standard operators are element-wise operations. When both `A` and `B` are matrices, `A * B` in Julia performs matrix multiplication, not element-wise multiplication as in Python. `A * B` in Julia is equivalent with `A @ B` in Python, whereas A * B in Python is equivalent with `A .* B` in Julia.

Reviewer: ```suggestion
  * In Julia, the standard operators over a matrix type are matrix operations, whereas, in Python, the standard operators are element-wise operations. When both `A` and `B` are matrices, `A * B` in Julia performs matrix multiplication, not element-wise multiplication as in Python. `A * B` in Julia is equivalent with `A @ B` in Python, whereas `A * B` in Python is equivalent with `A .* B` in Julia.
```

>   * The transpose operator `'` in Julia returns an adjoint of a vector (a lazy representation of row vector), whereas the transpose operator `.T` over a vector in Python returns the original vector (non-op).

Reviewer 6: I would refer to `'`  as the adjoint operator

Author: Thanks for the finding, is corrected

  * Functions in Python have one implementation (no polymorphism), whereas in Julia a function may contain multiple concrete implementations (called *Methods*), selected via multiple dispatch.
  * There are no classes in Julia. Instead they are structures (mutable or immutable), containing data but no methods. 
  * Calling a method of a class in Python (`a = MyClass(x), x.func(y)`) corresponds to a function call in Julia, e.g. `a = MyStruct(x), func(x::MyStruct, y)`. In general, multiple dispatch is more flexible and powerful than the Python class system.
  * Python classes can inherit from one or more superclasses. Julia structures may have exactly one abstract supertype. 
  * The Python code structure is defined by directories (Packages) and files (Modules), whereas the logical Julia program structure (Packages and Modules) is independent of the file strucutre (`include` for additional files).

>   * Conditional expression in Python `1 if x > 0 else -1` corresponds to the Ternary operator `x > 0 ? 1 : -1` in Julia.

Reviewer 2: As @waldyrious mentioned https://github.com/JuliaLang/julia/pull/35268#discussion_r399700492 I think it makes sense to lowercase "Ternary"

```suggestion
  * Conditional expression in Python `1 if x > 0 else -1` corresponds to the ternary operator `x > 0 ? 1 : -1` in Julia.
```

>   * Conditional assignment in Python `x = 1 if y>0 else -1` corresponds to the Ternary operator `x = y>0 ? 1 : -1` in Julia.

Reviewer 2: ```suggestion
  * Conditional expression in Python `1 if x > 0 else -1` corresponds to the Ternary operator `x > 0 ? 1 : -1` in Julia.
```

Reviewer 4: I'd suggest also lowercasing "Ternary".

>   * Macros in Julia can do what decorators in Python do, and much more.

Reviewer 2: I think this is a harmful statement to make casual Python programmers think that macros and decorators are somewhat similar.  If we are to include this, I think we need to make it clear that macros operate on expression and decorators operate on function objects.  Maybe what you want to say is "decorator-looking things" like `@inline` in Julia are actual macros?

Reviewer 2: The decorator syntax is much closer to `do` block in Julia actually.

Author: I had something like the `@time` macro in mind, which you could do with a decorator to the timed function in Python. But I agree that this may be misleading.
The suggestion to point out the different meaning of `@` is good, I'll change this point.

  * Exception handling in Julia is done using `try` - `catch` - `finally`, instead of `try` - `except` - `finally`. In contrast to Python, it is not recommended to use exception handling as part of the normal workflow in Julia due to performance reasons.
  * In Julia loops are fast, there is no need to write "vectorized" code for performance reasons.

>   * Non-constant global variables impose a significant performance loss in Julia and should therefore be avoided (see [Performance Tips](@ref man-performance-tips)). For Python, global variables are not associated with a performance penalty.  

Reviewer 2: Technically, globals are slow in Python, too...

```python
In [1]: def sum1(xs):
   ...:     acc = 0
   ...:     for x in xs:
   ...:         acc += x
   ...:     return acc

In [2]: def sum2(xs):
   ...:     global acc
   ...:     acc = 0
   ...:     for x in xs:
   ...:         acc += x
   ...:     return acc

In [3]: xs = range(1000)

In [4]: %timeit sum1(xs)
39.8 µs ± 399 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [5]: %timeit sum2(xs)
59.3 µs ± 778 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

Reviewer 3: So this claim is wrong since there is a performance penalty for globals in Python, it's just that locals are almost as slow as globals, so there's not as much of a penalty. Not sure how to word that best.

Author: Interesting, I was not aware of this speed difference in Python. Do you know the reason (type inference is required in any case)?

Suggestion for the text:

> For Python, global variables are associated with a much smaller performance penalty.

Reviewer 3: Same as in Julia, presumably: the Python interpreter knows that a local can't be changed out from under it, allowing it to do certain work once rather than every time it accesses the variable.

Reviewer 2: > Suggestion for the text:
> 
> > For Python, global variables are associated with a much smaller performance penalty.

I guess it's more like you don't care the performance penalty of globals in Python because the loops are slow in the first place?  Maybe the main message here is something like this?

> Be careful with non-constant global variables in Julia, especially in tight loops. Since you can write close-to-metal code in Julia (unlike Python), the effect of globals can be drastic (see Performance Tips).

Author: @tkf I like your suggestion, changing it accordingly.

>   * In Python `0`, `None`, empty list are evaluated to logical `False`. In Julia, you need explicit conversion to Bool.

Reviewer 2: It's strange to only talk about `False` in Python. Many/most things act like `True` as well.  Maybe something like

```suggestion
  * In Python, the majority of values can be used in logical context. In Julia, you need explicit conversion to Bool.
```

Author: I agree that your suggestion is better, especially since `[] == False` is `False` in Python.

Reviewer 4: I would leave an example or two, though, just to make sure it is clear what "used in logical context" means.

>   * In Python, the majority of values can be used in logical context. In Julia, you need explicit conversion to Bool.

Reviewer 2: re https://github.com/JuliaLang/julia/pull/35268#discussion_r399700341

@waldyrious Were you suggesting something like this?

```suggestion
  * In Python, the majority of values can be used in logical contexts (e.g, `1 if "a" else 0` evaluates to `1`). In Julia, you need explicit conversion to `Bool` (e.g., `"a" ? 1 : 0` throws).
```

@lungben What do you think?

Author: Yes, this example is helpful here.

Reviewer 4: I'm not sure relying on another language difference to illustrate this would be a good idea. How about simply mentioning that `0` and `None` are evaluated as logical `False` and that conversely, any/most things, such as a simple `"a"`, act like `True`? I can't think of a compact way to say that but thought I'd leave the suggestion.

Author: Maybe use a simple `if` statement for illustration?

> In Python, the majority of values can be used in logical contexts (e.g,  for `if "a":` the following block is executed, for `if "":` not). In Julia, you need explicit conversion to `Bool` (e.g., `if "a"` throws an exception).

Reviewer 4: Something like that sounds good, yeah :)

Reviewer 2: Sounds good to me, too.

Author: Thanks, is updated.

>   * In Python, the majority of values can be used in logical contexts (e.g, for `if "a":` the following block is executed, for `if "":` not). In Julia, you need explicit conversion to `Bool` (e.g. `if "a"` throws an exception).

Reviewer 4: The wording here seems slightly off, but I'm not quite sure how to improve it 🤔

How about this?

```suggestion
  * In Python, the majority of values can be used in logical contexts (e.g. `if "a":` means the following block is executed, and `if "":` means it is not). In Julia, you need explicit conversion to `Bool` (e.g. `if "a"` throws an exception).
```

>   * In Julia, a new local scope is introduced by most code blocks, including loops, `try`-`catch`-`finally` and comprehensions.

Reviewer: Maybe you could add that `if` blocks do not introduce a new local scope.

Author: good point, added it

>   * In Julia, a new local scope is introduced by most code blocks, including loops, `try`-`catch`-`finally` and comprehensions (note that `if` blocks do not introduce a new local scope).

Reviewer 2: List and generator comprehensions introduce a scope in Python as well.  I think listing it in differences is strange.

Author: is corrected
